core-rust/core-api-server/src/core_api/conversions/lts.rs

```diff
@@ -54,117 +59,59 @@ pub fn to_api_lts_committed_transaction_outcome(
 }
 
 pub fn to_api_lts_fungible_balance_changes(
+    database: &StateManagerDatabase,
     context: &MappingContext,
-    total_fee: Decimal,
+    fee_payments: &IndexMap<NodeId, Decimal>,
     balance_changes: &IndexMap<GlobalAddress, IndexMap<ResourceAddress, BalanceChange>>,
 ) -> Result<Vec<models::LtsEntityFungibleBalanceChanges>, MappingError> {
-    // TODO - until we have the proper information from the engine, we need to do some guessing here about which entities
-    // actually paid the fee. Ideally our guessing will align with 99% of transactions - and we can make clear in the docs
-    // it's not fully accurate for now but will be for launch.
-
-    // For now let's assume: The fee is paid by a single entity, and that entity is either:
-    // - The first entity to be debitted the exact fee, or if that doesn't exist:
-    // - The entity with the largest XRD debit.
-
-    let total_fee_balance_change = -total_fee;
-
-    let mut net_xrd_balance_change = Decimal::ZERO;
-    let mut exact_fee_debit = Option::<GlobalAddress>::None;
-    let mut biggest_xrd_debit = Option::<(Decimal, GlobalAddress)>::None;
-
-    for (entity_address, resource_changes) in balance_changes.iter() {
-        for (resource_address, balance_change) in resource_changes {
-            if *resource_address == RADIX_TOKEN {
-                let balance_change = get_fungible_balance(balance_change).unwrap();
-                if balance_change == total_fee_balance_change && exact_fee_debit.is_none() {
-                    exact_fee_debit = Some(*entity_address);
-                }
-                if biggest_xrd_debit.is_none() || balance_change < biggest_xrd_debit.unwrap().0 {
-                    biggest_xrd_debit = Some((balance_change, *entity_address));
-                }
-                net_xrd_balance_change += balance_change;
-            }
-        }
+    let mut fee_balance_changes = index_map_new();
+    let records = database.batch_get_ancestry(fee_payments.keys());
+    for (paid_fee_amount_xrd, ancestry) in fee_payments.values().zip(records) {
+        let ancestry = ancestry.expect("a vault must be owned by an account");
```

Comment:
We typically use `MappingError` instead of panics here, so the error gets turned into a 500. (Because a Core API issue/invariant is imo not sufficiently fatal to take down a node).

Reply:
Also, the expect message is incorrect - "a vault must be owned" - but not necessarily by an account! It could be some other global entity.